### DISCUSSION TITLE
Autocomplete property names

### DISCUSSION THREAD

**Discussion by stutrek** [eclipse-langium/langium#1672] [Q&A]
For eraser.io, we have a limited number of icons and colors in our DSL, and we want them to autocomplete.

```
Node [icon: globe, color: blue]
```

The `CompletionProvider` seems like a natural place to look, but the methods are either quite hard to override because they use utils that aren't export, or are not even being called when the cursor is in brackets for settings.

For our old, non-Langium system we're using some custom completion providers passed into Monaco.

Is there a way for us to tie autocomplete into settings?

**Comment by msujew** [comment]
Hey @stutrek 

> the methods are either quite hard to override because they use utils that aren't export

The utils are exported, but as dedicated namespaces. I.e. all functions from `utils/grammar-utils.ts` are exported as `GrammarUtils.<functionName>`.

> For our old, non-Langium system we're using some custom completion providers passed into Monaco.

Note that you can always write your own `CompletionProvider` that handles any logic you might want to implement. The default completion provider is just a default that works well for most languages and has dedicated template methods to adjust some behavior.

> Is there a way for us to tie autocomplete into settings?

You can inject the `ConfigurationProvider` that stores all vscode settings related to your language. Based on that, you can filter/adjusts the result of the normal completion provider. Do you have more context on that?
